bdk_core/src/tx_graph.rs

```
@@ -182,11 +189,110 @@ impl TxGraph {
+                        let serialized_txout = encode::serialize(output);
+                        let idx = (*idx as u32).to_be_bytes().to_vec();
+                        // add length of new item (idx + output)
+                        buf.extend(((idx.len() + serialized_txout.len()) as u32).to_be_bytes());
+                        // add index
+                        buf.extend(idx);
+                        // add output
+                        buf.extend(serialized_txout);
+                    }
+                }
+            }
+        }
+        let _size = writer.write(&buf)?;
+        Ok(())
+    }
+
+    pub fn from_db<T: Read + Seek>(
+        &mut self,
+        reader: &mut T,
+    ) -> Result<HashMap<Txid, TxNode>, Box<dyn Error>> {
+        let bytes = reader.bytes().collect::<io::Result<Vec<u8>>>()?;
+        let mut cursor = io::Cursor::new(bytes);
+
+        // skip the state_id + version bytes
+        cursor.set_position(8);
+
+        //while we haven't reached the end of the bytes iterator
+        // we want to read the first/length byte, read the type byte,
+        let mut length_bytes = [0u8; 4];
+        let mut tx_nodes = HashMap::new();
+        while cursor.read(&mut length_bytes)? != 0 {
+            let mut kind_byte = [0u8];
+            let _size = cursor.read(&mut kind_byte)?;
+
+            if kind_byte[0] == 1 {
+                // if the kind byte is full, read the next (length amount of) bytes,
+                // consensus deserialize that to full transaction
+                let length = u32::from_be_bytes(length_bytes);
+                //let mut tx_data = [0u8, length];
+                let mut tx_data = Vec::with_capacity(length as usize);
+                let _size = cursor.read(&mut tx_data)?;
+                let tx: Transaction = encode::deserialize(&tx_data)?;
+                let _inserted_tx = tx_nodes.insert(tx.txid(), TxNode::Whole(tx));
+            } else if kind_byte[0] == 2 {
```

`kind_byte` is not either of these shouldn't that return an error?

---

Yes, you are right. We should return an error here.